File: VCSJones.FiddlerCertGen/Interop/NCryptStorageProvider.cs
using System;
using System.ComponentModel;

namespace VCSJones.FiddlerCertGen.Interop
{
    public sealed class NCryptStorageProvider : IDisposable
    {
        private readonly NCryptStorageProviderSafeHandle _handle;

        public static NCryptStorageProvider MicrosoftSoftwareKeyStorageProvider { get; } = new NCryptStorageProvider("Microsoft Software Key Storage Provider");
        public static NCryptStorageProvider MicrosoftSmartCardProvider { get; } = new NCryptStorageProvider("Microsoft Smart Card Key Storage Provider");

        public string Name => NCryptPropertyReader.ReadStringUni(_handle, CngProperties.NCRYPT_NAME_PROPERTY);

        private NCryptStorageProvider(string name)
        {
            var result = NCrypt.NCryptOpenStorageProvider(out _handle, name, 0u);
            if (result != SECURITY_STATUS.ERROR_SUCCESS)
            {
                throw new Win32Exception(unchecked((int) result), "Failed to open the CNG storage provider.");
            }
        }



        internal NCryptStorageProviderSafeHandle Handle => _handle;

        public void Dispose()
        {
            _handle.Dispose();
        }
    }
}

File: VCSJones.FidderCertGenFS/NCryptStorageProvider.fs
namespace VCSJones.FiddlerCertGen

[<SealedAttribute>]
type NCryptStorageProvider(handle : NCryptStorageProviderSafeHandle) =
    static let softwareKeyStorageProvider = new NCryptStorageProvider("Microsoft Software Key Storage Provider")
    static let smartcardStorageProvider = new NCryptStorageProvider("Microsoft Smart Card Key Storage Provider")

    new(name : string) = 
        let mutable handle : NCryptStorageProviderSafeHandle = NCryptStorageProviderSafeHandle.Null
        let result = ExternInterop.NCryptOpenStorageProvider(&handle, name, 0u)
        if result <> 0u then
            invalidOp("Failed to open storage provider.")
        new NCryptStorageProvider(handle)

    static member MicrosoftSoftwareKeyStorageProvider = softwareKeyStorageProvider
    static member MicrosoftSmartCardProvider = smartcardStorageProvider

    member this.Handle = handle
    member this.Name = NCryptProperties.ReadStringUnicode handle CngConstants.NCRYPT_NAME_PROPERTY
        

    interface System.IDisposable with
        member this.Dispose() = handle.Dispose()